pandas/core/frame.py

```diff
@@ -3831,11 +3831,8 @@ def _getitem_multilevel(self, key):
                 result = self.reindex(columns=new_columns)
                 result.columns = result_columns
             else:
-                new_values = self.values[:, loc]
-                result = self._constructor(
-                    new_values, index=self.index, columns=result_columns
-                )
-                result = result.__finalize__(self)
+                result = self.iloc[:, loc]
+                result.columns = result_columns
```

Comment:
If we do away with going through `self.values` (which is a good catch! that probably stems from the time we only had consolidated numpy dtypes, so whenever you had a single dtype we assumed a single numpy array), we might as well combine the `if self._is_mixed_type: .. else: ..` paths? I am not sure there is any benefit over using iloc vs reindex.

Reply:
Having a look at this locally, it seems like these separate paths are now redundant and the reindex block can be used in both. 
(There was not deliberate choice in my original change to use iloc instead of reindex, I am just less familiar with reindex). I'm also unfortunately ignorant of how they compare performance wise.

Reply:
They should be equivalent, I think (especially since we afterwards still set the resulting column names), generally they should end up using the same code under the hood. Since the existing code was using reindex, I think it's fine to continue using that.